### PR TITLE
Add a metric for counting the number of suppressed spans.

### PR DESCRIPTION
**Pull request by jkwatson**
This will help in diagnosing mysterious "missing span" issues.

### CODE REVIEW

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 import org.slf4j.LoggerFactory;
 
 public abstract class BaseTracer {
-  private static final Logger log = LoggerFactory.getLogger(HttpServerTracer.class);
+  private static final Logger log = LoggerFactory.getLogger(BaseTracer.class);
```
Comment (jkwatson) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java:28]: heh. this was also wrong and very confusing when I saw it in the logs while verifying. :)

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java [hunk 2]
```diff
@@ -81,14 +81,27 @@ public Span getCurrentSpan() {
...
-        return !inClientSpan(context);
+        result = !inClientSpan(context);
+        break;
       case SERVER:
-        return !inServerSpan(context);
+        result = !inServerSpan(context);
+        break;
       default:
-        return true;
+        result = true;
```
Comment (breedx-splk) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java:93]: you could move this assignment to the declaration and remove the default to save a couple lines of code.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java [hunk 3]
```diff
@@ -81,14 +81,27 @@ public Span getCurrentSpan() {
...
     }
+    if (!result) {
+      if (log.isDebugEnabled()) {
```
Comment (breedx-splk) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java:96]: prefer to roll these nested `if`s to an `&&`.

**Review by breedx-splk — APPROVED** [review]
Looks good to me, a few small things for polish if you wanna.

**Review by trask — COMMENTED** [review]
Oh, I just remembered, we have been removing CallDepth tracking (which prevents an instrumentation from creating multiple nested spans) in places where it is redundant with the nested CLIENT span check (https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/907), but that means those instrumentations will trigger this "suppressed instrumentation" log even though that's not really the case.

I'm not really sure how to detect a "real" suppressed instrumentation 🤔

**Comment by jkwatson** [comment]
> Oh, I just remembered, we have been removing CallDepth tracking (which prevents an instrumentation from creating multiple nested spans) in places where it is redundant with the nested CLIENT span check (#907), but that means those instrumentations will trigger this "suppressed instrumentation" log even though that's not really the case.
> 
> I'm not really sure how to detect a "real" suppressed instrumentation 🤔

Can you point me at an example of this? Maybe we need to update the code to only use this code path when we're actually doing "real" span suppression (whatever that means).

**Comment by trask** [comment]
Check out #1811 for a bunch of places where I removed the CallDepth tracking since we can now rely on nested client span suppression.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java [hunk 4]
```diff
@@ -81,14 +81,27 @@ public Span getCurrentSpan() {
   }
 
   protected final boolean shouldStartSpan(Kind proposedKind, Context context) {
+    boolean result;
```
Comment (iNikem) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java:84]: Rename `result` to `suppressed` and remove some of negations?

**Review by iNikem — APPROVED** [review]

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java (hunk 4 above)
Comment (jkwatson) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java]: done

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java (hunk 3 above)
Comment (jkwatson) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java]: done

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java (hunk 2 above)
Comment (jkwatson) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java]: done

**Comment by jkwatson** [comment]
> Oh, I just remembered, we have been removing CallDepth tracking (which prevents an instrumentation from creating multiple nested spans) in places where it is redundant with the nested CLIENT span check (#907), but that means those instrumentations will trigger this "suppressed instrumentation" log even though that's not really the case.
> 
> I'm not really sure how to detect a "real" suppressed instrumentation 🤔

So... is this an objection to this PR? Or just a musing that the word "suppression" might not be 100% accurate? If we're suppressing spans, we're suppressing them, whether or not it's due to too many nested spans, or other reasons, no?

**Comment by trask** [comment]
> So... is this an objection to this PR? Or just a musing that the word "suppression" might not be 100% accurate? If we're suppressing spans, we're suppressing them, whether or not it's due to too many nested spans, or other reasons, no?

I'm not sure how it's going to be useful since it's going to log lots of false positives due to the removal of CallDepth tracking.

Interestingly, in order to implement https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/465#issuecomment-769429325, we will probably have to bring back CallDepth tracking (or something similar), so we know whether we should suppress due to "CallDepth", or whether we should add an event to current span.

**Comment by jkwatson** [comment]
> > So... is this an objection to this PR? Or just a musing that the word "suppression" might not be 100% accurate? If we're suppressing spans, we're suppressing them, whether or not it's due to too many nested spans, or other reasons, no?
> 
> I'm not sure how it's going to be useful since it's going to log lots of false positives due to the removal of CallDepth tracking.
> 
> Interestingly, in order to implement [#465 (comment)](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/465#issuecomment-769429325), we will probably have to bring back CallDepth tracking (or something similar), so we know whether we should suppress due to "CallDepth", or whether we should add an event to current span.

Are those actually false positives though? We are suppressing the spans in those cases, aren't we?

**Comment by jkwatson** [comment]
I changed this to use a LongCounter to keep track of the suppressed spans. There is one optimization to be made if we want to go this route.

**Comment by jkwatson** [comment]
@trask @iNikem @anuraaga thoughts on the current state, using metrics for this?

**Review by trask — APPROVED** [review]
> thoughts on the current state, using metrics for this?

I think it's a cool idea, let's try it out 👍

**Review by anuraaga — APPROVED** [review]

**Comment by iNikem** [comment]
Metrics are cool, but... :) They are alpha and will remain such for a long time. Do we want for `instrumentation-api` do depend on alpha artifact? We cannot then release this API as stable., can we?

**Comment by anuraaga** [comment]
@iNikem Hmm this is a good point. Even if it's an implementation detail for us, it could still cause dependency hell for users regardless.

But my guess is metrics will stabilize before instrumentation - we still have the question of naming things tracer vs instrumenter, only because of metrics.

So I'm ok with it and if we actually release instrumentation first at the time we can flag guard or something similar.

**Comment by iNikem** [comment]
> But my guess is metrics will stabilize before instrumentation

You are an optimist about metrics and pessimist about instrumentations :D

**Comment by jkwatson** [comment]
Just thinking out loud, but another option would be to create some sort of "Supportability" interface that could be implemented by by whatever the user would prefer (logging, or metrics, or micrometer metrics, or a no-op). So, we could decouple the implementation from the need for providing some sort of observability for the internals of the instrumentation APIs/agent.

**Review by iNikem — APPROVED** [review]

**Pull request merged**